File: apps/frontend/lib/api/roles.ts
/**
 * API client functions for role management
 */
import apiClient from './client';
import { 
  Permission, 
  PermissionParams, 
  Role, 
  RoleParams, 
  RoleAssignmentParams,
  User,
  UserRoleUpdateParams,
  SystemFeaturesResponse
} from './types';

// Base endpoints
const ROLES_ENDPOINT = '/roles';
const PERMISSIONS_ENDPOINT = '/permissions';
const ADMIN_ENDPOINT = '/admin';
const USERS_ENDPOINT = '/users';
const SYSTEM_ENDPOINT = '/system';

/**
 * Get all roles
 */
export async function getRoles(): Promise<Role[]> {
  const response = await apiClient.get(ROLES_ENDPOINT);
  return response.data;
}

/**
 * Get a specific role by ID
 */
export async function getRole(id: number): Promise<Role> {
  const response = await apiClient.get(`${ROLES_ENDPOINT}/${id}`);
  return response.data;
}

/**
 * Create a new role
 */
export async function createRole(role: RoleParams): Promise<Role> {
  const response = await apiClient.post(ROLES_ENDPOINT, role);
  return response.data;
}

/**
 * Update an existing role
 */
export async function updateRole(id: number, role: RoleParams): Promise<Role> {
  const response = await apiClient.put(`${ROLES_ENDPOINT}/${id}`, role);
  return response.data;
}

/**
 * Delete a role
 */
export async function deleteRole(id: number): Promise<void> {
  await apiClient.delete(`${ROLES_ENDPOINT}/${id}`);
}

/**
 * Get all permissions
 */
export async function getPermissions(): Promise<Permission[]> {
  const response = await apiClient.get(PERMISSIONS_ENDPOINT);
  return response.data;
}

/**
 * Get a specific permission by ID
 */
export async function getPermission(id: number): Promise<Permission> {
  const response = await apiClient.get(`${PERMISSIONS_ENDPOINT}/${id}`);
  return response.data;
}

/**
 * Create a new permission
 */
export async function createPermission(permission: PermissionParams): Promise<Permission> {
  const response = await apiClient.post(PERMISSIONS_ENDPOINT, permission);
  return response.data;
}

/**
 * Update an existing permission
 */
export async function updatePermission(id: number, permission: PermissionParams): Promise<Permission> {
  const response = await apiClient.put(`${PERMISSIONS_ENDPOINT}/${id}`, permission);
  return response.data;
}

/**
 * Delete a permission
 */
export async function deletePermission(id: number): Promise<void> {
  await apiClient.delete(`${PERMISSIONS_ENDPOINT}/${id}`);
}

/**
 * Assign a role to a user
 */
export async function assignRoleToUser(roleId: number, userId: number): Promise<void> {
  await apiClient.post(`${ROLES_ENDPOINT}/${roleId}/users/${userId}`);
}

/**
 * Remove a role from a user
 */
export async function removeRoleFromUser(roleId: number, userId: number): Promise<void> {
  await apiClient.delete(`${ROLES_ENDPOINT}/${roleId}/users/${userId}`);
}

/**
 * Update a user's legacy role (admin/user)
 */
export async function updateUserRole(userId: number, roleData: UserRoleUpdateParams): Promise<{ message: string }> {
  const response = await apiClient.put(`${ADMIN_ENDPOINT}/users/${userId}/role`, roleData);
  return response.data;
}

/**
 * Make the current user an admin (special setup endpoint)
 */
export async function makeSelfAdmin(): Promise<{ message: string }> {
  const response = await apiClient.put(`${ADMIN_ENDPOINT}/self/make-admin`);
  return response.data;
}

/**
 * Get all users
 */
export async function getUsers(): Promise<User[]> {
  const response = await apiClient.get(USERS_ENDPOINT);
  return response.data;
}

/**
 * Get a specific user by ID
 */
export async function getUser(id: number): Promise<User> {
  const response = await apiClient.get(`${USERS_ENDPOINT}/${id}`);
  return response.data;
}

/**
 * Get system features status (admin only)
 */
export async function getSystemFeatures(): Promise<SystemFeaturesResponse> {
  const response = await apiClient.get(`${SYSTEM_ENDPOINT}/features`);
  return response.data;
}

File: apps/backend/src/routes/roles.py
"""
Role management routes for the JurisAI API.
"""

from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.orm import Session

from src.core.database import get_db
from src.models.role import Role
from src.models.permission import Permission
from src.models.user import User
from src.routes.auth import get_current_user

# Create router
router = APIRouter(prefix="/auth/roles", tags=["roles"])

# Models
class PermissionResponse(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    resource: str
    action: str

    class Config:
        orm_mode = True


class RoleBase(BaseModel):
    name: str
    description: Optional[str] = None
    is_default: Optional[bool] = False


class RoleCreate(RoleBase):
    permission_ids: List[int] = []


class RoleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_default: Optional[bool] = None
    permission_ids: Optional[List[int]] = None


class RoleResponse(RoleBase):
    id: int
    created_at: datetime
    updated_at: datetime
    permissions: List[PermissionResponse] = []

    class Config:
        orm_mode = True


# Helper function to check if user is admin
def is_admin(user: User) -> bool:
    """Check if user is an admin."""
    if user.role == "admin":
        return True
    
    for role in user.roles:
        if role.name == "admin":
            return True
    
    return False


# Routes
@router.get("/", response_model=List[RoleResponse])
async def get_roles(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get all roles."""
    # Check if user is admin or has permission to view roles
    if not current_user.has_permission("role", "read") and not is_admin(current_user):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions",
        )
    
    roles = db.query(Role).offset(skip).limit(limit).all()
    return roles


@router.get("/{role_id}", response_model=RoleResponse)
async def get_role(
    role_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get a specific role by ID."""
    # Check if user is admin or has permission to view roles
    if not current_user.has_permission("role", "read") and not is_admin(current_user):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions",
        )
    
    role = db.query(Role).filter(Role.id == role_id).first()
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found",
        )
    
    return role


@router.post("/", response_model=RoleResponse, status_code=status.HTTP_201_CREATED)
async def create_role(
    role_data: RoleCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a new role."""
    # Check if user is admin or has permission to create roles
    if not current_user.has_permission("role", "create") and not is_admin(current_user):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions",
        )
    
    # Check if role with this name already exists
    existing_role = db.query(Role).filter(Role.name == role_data.name).first()
    if existing_role:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Role with this name already exists",
        )
    
    # Convert is_default bool to int
    is_default_int = 1 if role_data.is_default else 0
    
    # Create new role
    new_role = Role(
        name=role_data.name,
        description=role_data.description,
        is_default=is_default_int,
    )
    
    # Add permissions if provided
    if role_data.permission_ids:
        permissions = db.query(Permission).filter(
            Permission.id.in_(role_data.permission_ids)
        ).all()
        
        if len(permissions) != len(role_data.permission_ids):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="One or more permission IDs are invalid",
            )
        
        new_role.permissions = permissions
    
    db.add(new_role)
    db.commit()
    db.refresh(new_role)
    
    return new_role


@router.put("/{role_id}", response_model=RoleResponse)
async def update_role(
    role_id: int,
    role_data: RoleUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update a role."""
    # Check if user is admin or has permission to update roles
    if not current_user.has_permission("role", "update") and not is_admin(current_user):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions",
        )
    
    # Get role
    role = db.query(Role).filter(Role.id == role_id).first()
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found",
        )
    
    # Update role fields if provided
    if role_data.name is not None:
        # Check if another role with this name exists
        existing_role = db.query(Role).filter(
            Role.name == role_data.name,
            Role.id != role_id,
        ).first()
        
        if existing_role:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Another role with this name already exists",
            )
        
        role.name = role_data.name
    
    if role_data.description is not None:
        role.description = role_data.description
    
    if role_data.is_default is not None:
        role.is_default = 1 if role_data.is_default else 0
    
    # Update permissions if provided
    if role_data.permission_ids is not None:
        permissions = db.query(Permission).filter(
            Permission.id.in_(role_data.permission_ids)
        ).all()
        
        if len(permissions) != len(role_data.permission_ids):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="One or more permission IDs are invalid",
            )
        
        role.permissions = permissions
    
    role.updated_at = datetime.now()
    db.commit()
    db.refresh(role)
    
    return role


@router.delete("/{role_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_role(
    role_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Delete a role."""
    # Check if user is admin or has permission to delete roles
    if not current_user.has_permission("role", "delete") and not is_admin(current_user):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions",
        )
    
    # Get role
    role = db.query(Role).filter(Role.id == role_id).first()
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found",
        )
    
    # Delete role
    db.delete(role)
    db.commit()
    
    return None


@router.post("/{role_id}/users/{user_id}", status_code=status.HTTP_204_NO_CONTENT)
async def assign_role_to_user(
    role_id: int,
    user_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Assign a role to a user."""
    # Check if user is admin or has permission to update users
    if not current_user.has_permission("user", "update") and not is_admin(current_user):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions",
        )
    
    # Get role and user
    role = db.query(Role).filter(Role.id == role_id).first()
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found",
        )
    
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found",
        )
    
    # Check if user already has this role
    if role in user.roles:
        return None  # No need to add it again
    
    # Assign role to user
    user.roles.append(role)
    db.commit()
    
    return None


@router.delete("/{role_id}/users/{user_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_role_from_user(
    role_id: int,
    user_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Remove a role from a user."""
    # Check if user is admin or has permission to update users
    if not current_user.has_permission("user", "update") and not is_admin(current_user):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Insufficient permissions",
        )
    
    # Get role and user
    role = db.query(Role).filter(Role.id == role_id).first()
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found",
        )
    
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found",
        )
    
    # Check if user has this role
    if role not in user.roles:
        return None  # No need to remove it
    
    # Remove role from user
    user.roles.remove(role)
    db.commit()
    
    return None
